In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import scipy.stats as stats
import timeit

%matplotlib inline


import warnings 

pd.set_option('display.max_rows', 1000)
pd.set_option('display.max_columns', 500)
pd.set_option('display.width', 1000)

warnings.filterwarnings('ignore')


In [2]:
# Replace the path with the correct path for your data.
y2015 = pd.read_csv(
    'https://www.dropbox.com/s/0so14yudedjmm5m/LoanStats3d.csv?dl=1',
    skipinitialspace=True,
    header=1
)

# Note the warning about dtypes.

In [3]:
# Convert ID and Interest Rate to numeric.
y2015['id'] = pd.to_numeric(y2015['id'], errors='coerce')
y2015['int_rate'] = pd.to_numeric(y2015['int_rate'].str.strip('%'), errors='coerce')

# Drop other columns with many unique variables
y2015.drop(['url', 'emp_title', 'zip_code', 'earliest_cr_line', 'revol_util',
            'sub_grade', 'addr_state', 'desc','policy_code'], 1, inplace=True)

#'policy_code'

In [12]:
# Remove two summary rows at the end that don't actually contain data.
y2015 = y2015[:-2]
y2015.fillna(0)

id   member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  int_rate  installment grade emp_length home_ownership  annual_inc verification_status   issue_d  loan_status pymnt_plan             purpose                    title    dti  delinq_2yrs  inq_last_6mths  mths_since_last_delinq  mths_since_last_record  open_acc  pub_rec  revol_bal  total_acc initial_list_status  out_prncp  out_prncp_inv  total_pymnt  total_pymnt_inv  total_rec_prncp  total_rec_int  total_rec_late_fee  recoveries  collection_recovery_fee last_pymnt_d  last_pymnt_amnt next_pymnt_d last_credit_pull_d  collections_12_mths_ex_med  mths_since_last_major_derog application_type  annual_inc_joint  dti_joint verification_status_joint  acc_now_delinq  tot_coll_amt  tot_cur_bal  open_acc_6m  open_il_6m  open_il_12m  open_il_24m  mths_since_rcnt_il  total_bal_il  il_util  open_rv_12m  open_rv_24m  max_bal_bc  all_util  total_rev_hi_lim  inq_fi  total_cu_tl  inq_last_12m  acc_open_past_24mths  \
0       68009401.0  72868139.0    16000.0      16000.0          16000.0   60 months     14.85       379.39     C  10+ years       MORTGAGE     48000.0        Not Verified  Dec-2015      Current          n         credit_card  Credit card refinancing  33.18          0.0             0.0                    33.0                     2.0      11.0      2.0    19108.0       19.0                   w   13668.88       13668.88      4519.68          4519.68          2331.12        2188.56                 0.0         0.0                      0.0     Jan-2017           379.39     Jan-2017           Jan-2017                         0.0                         33.0       INDIVIDUAL               0.0        0.0                         0             0.0           0.0      31329.0          1.0         2.0          1.0          2.0                 6.0       12221.0     91.0          1.0          4.0     14982.0      65.0          284700.0     0.0          0.0           1.0                   6.0   
1       68354783.0  73244544.0     9600.0       9600.0           9600.0   36 months      7.49       298.58     A    8 years       MORTGAGE     60000.0        Not Verified  Dec-2015      Current          n         credit_card  Credit card refinancing  22.44          0.0             0.0                     0.0                     0.0       7.0      0.0     7722.0        9.0                   w    6635.69        6635.69      3572.97          3572.97          2964.31         608.66                 0.0         0.0                      0.0     Jan-2017           298.58     Jan-2017           Jan-2017                         0.0                          0.0       INDIVIDUAL               0.0        0.0                         0             0.0           0.0      55387.0          0.0         3.0          0.0          0.0                47.0       47665.0     43.0          2.0          2.0      4744.0      50.0           13000.0     0.0          0.0           2.0                   2.0   
2       68466916.0  73356753.0    25000.0      25000.0          25000.0   36 months      7.49       777.55     A  10+ years       MORTGAGE    109000.0        Not Verified  Dec-2015   Fully Paid          n  debt_consolidation       Debt consolidation  26.02          0.0             1.0                     0.0                     0.0       9.0      0.0    20862.0       19.0                   w       0.00           0.00     26224.23         26224.23         25000.00        1224.23                 0.0         0.0                      0.0     Sep-2016         20807.39            0           Jan-2017                         0.0                          0.0       INDIVIDUAL               0.0        0.0                         0             0.0           0.0     305781.0          0.0         3.0          0.0          1.0                13.0       47194.0     58.0          0.0          1.0      8937.0      57.0           38400.0     1.0          0.0           1.0                   2.0   
3       68466961.0  73356799.0    28000.

## run classifier to select Continuous Variables 

In [15]:
from sklearn import ensemble
from sklearn.model_selection import cross_val_score

rfc = ensemble.RandomForestClassifier()

#'grade','verification_status','last_pymnt_d','verification_status_joint','total_il_high_credit_limit'

X = y2015[[
'loan_amnt',
 'funded_amnt',
 'funded_amnt_inv',
 'int_rate',
 'installment',
 'annual_inc',
 'dti',
 'delinq_2yrs',
 'inq_last_6mths',
 'mths_since_last_delinq',
 'mths_since_last_record',
 'open_acc',
 'pub_rec',
 'revol_bal',
 'total_rec_int',
 'total_rec_late_fee',
 'recoveries',
 'collection_recovery_fee',
 'last_pymnt_amnt',
 'collections_12_mths_ex_med',
 'mths_since_last_major_derog',
 'annual_inc_joint',
 'dti_joint',
 'acc_now_delinq',
 'tot_coll_amt',
 'tot_cur_bal',
 'open_acc_6m',
 'open_il_6m',
 'open_il_12m',
 'open_il_24m',
 'mths_since_rcnt_il',
 'total_bal_il',
 'il_util',
 'open_rv_12m',
 'open_rv_24m',
 'max_bal_bc',
 'all_util',
 'total_rev_hi_lim',
 'inq_fi',
 'total_cu_tl',
 'inq_last_12m',
 'acc_open_past_24mths',
 'avg_cur_bal',
 'bc_open_to_buy',
 'bc_util',
 'chargeoff_within_12_mths',
 'delinq_amnt',
 'mo_sin_old_il_acct',
 'mo_sin_old_rev_tl_op',
 'mo_sin_rcnt_rev_tl_op',
 'mo_sin_rcnt_tl',
 'mort_acc',
 'mths_since_recent_bc',
 'mths_since_recent_bc_dlq',
 'mths_since_recent_inq',
 'mths_since_recent_revol_delinq',
 'num_accts_ever_120_pd',
 'num_actv_bc_tl',
 'num_actv_rev_tl',
 'num_bc_sats',
 'num_bc_tl',
 'num_il_tl',
 'num_op_rev_tl',
 'num_rev_accts',
 'num_rev_tl_bal_gt_0',
 'num_sats',
 'num_tl_120dpd_2m',
 'num_tl_30dpd',
 'num_tl_90g_dpd_24m',
 'num_tl_op_past_12m',
 'pct_tl_nvr_dlq',
 'percent_bc_gt_75',
 'pub_rec_bankruptcies',
 'tax_liens',
 'tot_hi_cred_lim',
 'total_bal_ex_mort',
 'total_bc_limit',
 'total_il_high_credit_limit']]








           
           #'last_pymnt_amnt','total_rec_prncp']]
#X = y2015[['loan_amnt','max_bal_bc']]

#X = y2015[['grade','verification_status','last_pymnt_d','verification_status_joint']]#,'loan_amnt']]
Y = y2015['loan_status']
X = pd.get_dummies(X)
X = X.dropna(axis=1)




[0.92517395 0.92173066 0.91719503 0.91802422 0.91861791 0.91935407
 0.91963713 0.9215351  0.92096326 0.92103263]


In [19]:
from sklearn.model_selection import train_test_split

X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2, random_state=42)


In [31]:
#from sklearn import ensemble
from sklearn.feature_selection import SelectFromModel
sel = SelectFromModel(ensemble.RandomForestClassifier(n_estimators = 100))
sel.fit(X_train, Y_train)

SelectFromModel(estimator=RandomForestClassifier(bootstrap=True,
                                                 class_weight=None,
                                                 criterion='gini',
                                                 max_depth=None,
                                                 max_features='auto',
                                                 max_leaf_nodes=None,
                                                 min_impurity_decrease=0.0,
                                                 min_impurity_split=None,
                                                 min_samples_leaf=1,
                                                 min_samples_split=2,
                                                 min_weight_fraction_leaf=0.0,
                                                 n_estimators=100, n_jobs=None,
                                                 oob_score=False,
                                                 random_state=None, verbose=0,
  

In [32]:
sel.get_support()


array([False, False, False,  True, False, False, False, False, False,
       False, False, False,  True, False,  True,  True,  True, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False])

In [33]:
selected_feat= X_train.columns[(sel.get_support())]
len(selected_feat)

5

In [34]:
print(selected_feat)

Index(['int_rate', 'total_rec_int', 'recoveries', 'collection_recovery_fee', 'last_pymnt_amnt'], dtype='object')


## Run Classifier to select catagorical variables 

In [42]:
y2015.select_dtypes(include=['object']).columns.tolist()

['term',
 'grade',
 'emp_length',
 'home_ownership',
 'verification_status',
 'issue_d',
 'loan_status',
 'pymnt_plan',
 'purpose',
 'title',
 'initial_list_status',
 'last_pymnt_d',
 'next_pymnt_d',
 'last_credit_pull_d',
 'application_type',
 'verification_status_joint']

In [ ]:
U = y2015[['term',
 'grade',
 'emp_length',
 'home_ownership',
 'verification_status',
 'issue_d',
 'pymnt_plan',
 'purpose',
 'title',
 'initial_list_status',
 'last_pymnt_d',
 'next_pymnt_d',
 'last_credit_pull_d',
 'application_type',
 'verification_status_joint']]
V = y2015['loan_status']
U = pd.get_dummies(U)
U = U.dropna(axis=1)


In [43]:
from sklearn.model_selection import train_test_split

U_train, U_test, V_train, V_test = train_test_split(U, V, test_size=0.2, random_state=42)


In [44]:
#from sklearn import ensemble
from sklearn.feature_selection import SelectFromModel
sel = SelectFromModel(ensemble.RandomForestClassifier(n_estimators = 100))
sel.fit(U_train, V_train)

SelectFromModel(estimator=RandomForestClassifier(bootstrap=True,
                                                 class_weight=None,
                                                 criterion='gini',
                                                 max_depth=None,
                                                 max_features='auto',
                                                 max_leaf_nodes=None,
                                                 min_impurity_decrease=0.0,
                                                 min_impurity_split=None,
                                                 min_samples_leaf=1,
                                                 min_samples_split=2,
                                                 min_weight_fraction_leaf=0.0,
                                                 n_estimators=100, n_jobs=None,
                                                 oob_score=False,
                                                 random_state=None, verbose=0,
  

In [45]:
sel.get_support()

array([ True,  True,  True,  True,  True, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False,  True, False, False, False, False,  True, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False])

In [47]:
selected_feat= U_train.columns[(sel.get_support())]
len(selected_feat)

7

In [48]:
print(selected_feat)

Index(['total_il_high_credit_limit', 'int_rate', 'total_rec_int', 'recoveries', 'collection_recovery_fee', 'last_pymnt_d_Dec-2016', 'last_pymnt_d_Jan-2017'], dtype='object')


In [65]:
rfc = ensemble.RandomForestClassifier()


I = y2015[['total_il_high_credit_limit', 'int_rate',
       'total_rec_int', 'recoveries',
       'collection_recovery_fee', 'last_pymnt_d',
       'int_rate', 'total_rec_int',
       'recoveries', 'collection_recovery_fee']]

#'last_pymnt_d',, 'last_pymnt_amnt'
J = y2015['loan_status']
I = pd.get_dummies(I)
I = I.dropna(axis=1)





In [66]:

#, 'last_pymnt_amnt'

score=cross_val_score(rfc, I, J, cv=10)
print(score)
#type(score[1])

[0.86407352 0.86046401 0.87027143 0.87810496 0.87456661 0.88544289
 0.88700753 0.89063836 0.88564847 0.81819693]


After looking through all the data. without using anything related to payment the best score is 0.818 and this isnt acounting for problems of class imbalence in the target variable. it appears that the incorperation of 